optimum/fx/parallelization/api.py

```diff
@@ -0,0 +1,155 @@
+    """
+    parallel_config = Config()
+    for k, v in kwargs.items():
+        if k in parallel_config.__dict__:
+            setattr(parallel_config, k, v)
+    kwargs = {k: v for k, v in kwargs.items() if k not in parallel_config.__dict__}
+
+    if isinstance(model, str):
+        from transformers import AutoConfig
+        from transformers.utils import SAFE_WEIGHTS_INDEX_NAME, WEIGHTS_INDEX_NAME
+
+        is_local = os.path.isdir(model)
+        allow_patterns = ["*.safetensors", "*.bin"]
+        if not is_local:
+            hf_folder = download_files_from_hf(
+                model_name_or_path=model,
+                cache_dir=cache_dir,
+                allow_patterns=allow_patterns,
+                revision=revision,
+                local_files_only=local_files_only,
+                skip_download_weights=skip_load_weights,
+            )
+        else:
+            hf_folder = model
+
+        # should be able to load config using only local files
+        model_config, kwargs = AutoConfig.from_pretrained(
+            hf_folder, revision=revision, local_files_only=True, return_unused_kwargs=True, **kwargs
+        )
+
+        # try getting model class info from config
+        model_arch = model_config.architectures
+        model_cls = getattr(importlib.import_module("transformers"), model_arch[0])
+
+        if not skip_load_weights:
+            use_safetensors = False
+            for pattern in allow_patterns:
+                if len(glob.glob(os.path.join(hf_folder, pattern))) > 0:
+                    use_safetensors = pattern == "*.safetensors"
+                    break
+            index_path = os.path.join(hf_folder, SAFE_WEIGHTS_INDEX_NAME if use_safetensors else WEIGHTS_INDEX_NAME)
+            if os.path.isfile(index_path):
+                with open(index_path) as f:
+                    index_dict = json.load(f)
+                parallel_ctx.weight_map = {k: os.path.join(hf_folder, v) for k, v in index_dict["weight_map"].items()}
+            weight_files = glob.glob(os.path.join(hf_folder, "*.safetensors" if use_safetensors else "*.bin"))
+            if not use_safetensors:
+                weight_map = parallel_ctx.weight_map if parallel_ctx.weight_map else {}
+                convert_bin_to_safetensors(model, cache_dir, weight_files, weight_map)
+                parallel_ctx.weight_map = weight_map
+
+            # try directly construct weight_map from weight files, should have safetensors file on disk in any case
+            if not parallel_ctx.weight_map:
+                from safetensors import safe_open
+
+                weight_map, weight_files = {}, glob.glob(os.path.join(hf_folder, "*.safetensors"))
+                for weight_file in weight_files:
+                    with safe_open(filename=weight_file, framework="pt") as f:
+                        for key in f.keys():
+                            weight_map[key] = weight_file
+                parallel_ctx.weight_map = weight_map
```

Comment:
I move the logic into `utils` so it looks cleaner in `api`, but the logic itself is indeed complex, because we need to take care of situations where a local directory is passed, then the only thing we can do is to try to peek inside the folder and see if there are safetensors/bin files, and if there are only bin files, we need to convert them into safetensors, and if there is an index file, we load the `weight_map` directly from it, otherwise we try scanning all the weight files in the folder and assemble a `weight_map` out of them.